### PR TITLE
Interfacing EvtCalc in FairShip

### PR DESCRIPTION
**Pull request by mferril**
### Vanilla implementation of the EventCalc LLPs decay event sampler in FairShip for inclusive final states.

Additions & modifications to master:
* FairShip/shipgen/EvtCalcGenerator(.cxx/.h)
* FairShip/macro/run_simScript.py
* FairShip/macro/convertEvtCalc.py

To do list (in order of priority):

- [ ] Including more LLP models ( Issue #536 )
- [ ] Incorporating the event sampler code within FairShip ( Issue #537 )
- [ ] Interfacing PYTHIA for hadronisation of final states with qq' ( Issue #538 )
- [ ] Adding features to produce exclusive decays ( Issue #539 )
- [ ] Implementing systematic uncertainties in the production and decay calculations ( Issue #540 )

FYI @maksymovchynnikov

### CODE REVIEW

**Comment by olantwin** [comment]
A preliminary remark and a question: 

* The history needs some cleaning up, we can do that once everything else is ready for review. Only the merge commit might be nice to remove already now.
* Which of the todo items would you consider necessary for the first version and which can we tackle later?

File: shipgen/EvtCalcGenerator.cxx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// -------------------------------------------------------------------------
+// -----   Default constructor   -------------------------------------------
+Bool_t EvtCalcGenerator::Init(const char* fileName) {
+  return Init(fileName, 0);
+}
+// -----   Default constructor   -------------------------------------------
+Bool_t EvtCalcGenerator::Init(const char* fileName, const int firstEvent) {
+  
+  cout << "Info EvtCalcGenerator: Opening input file " << fileName << endl;
+  fInputFile  = new TFile(fileName);
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:30]:
```suggestion
  fInputFile = TFile::Open(fileName, "read");
```

File: shipgen/EvtCalcGenerator.cxx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  LOGF(info, "Info EvtCalcGenerator: Opening input file %s", fileName);
+  if (fInputFile->IsZombie() or !fInputFile) {
+      LOG(FATAL) << "Info EvtCalcGenerator: Error opening input file";
+      return kFALSE; }
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:34]: Remove, redundant with TFile::Open

File: shipgen/EvtCalcGenerator.cxx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+// -------------------------------------------------------------------------
+
+
+// -----   Destructor   ----------------------------------------------------
+EvtCalcGenerator::~EvtCalcGenerator()
+{
+ // cout << "destroy Ntuple" << endl;
+ fInputFile->Close();
+ fInputFile->Delete();
+ delete fInputFile;
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:79]: This line is probably redundant.

File: shipgen/EvtCalcGenerator.cxx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Bool_t EvtCalcGenerator::Init(const char* fileName, const int firstEvent) {
+  
+  cout << "Info EvtCalcGenerator: Opening input file " << fileName << endl;
+  fInputFile  = new TFile(fileName);
+  LOGF(info, "Info EvtCalcGenerator: Opening input file %s", fileName);
+  if (fInputFile->IsZombie() or !fInputFile) {
+      LOG(FATAL) << "Info EvtCalcGenerator: Error opening input file";
+      return kFALSE; }
+  
+  fTree = (TTree *)fInputFile->Get("LLP_tree");
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:36]: Please don't use C-style casts

File: shipgen/EvtCalcGenerator.cxx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ fInputFile->Close();
+ fInputFile->Delete();
+ delete fInputFile;
+}
+// -------------------------------------------------------------------------
+
+// -----   Passing the event   ---------------------------------------------
+Bool_t EvtCalcGenerator::ReadEvent(FairPrimaryGenerator* cpg)
+{
+  if (fn==fNevents){ LOG(WARNING) << "End of input file. Rewind."; }
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:86]: Is the "rewind" implemented?

File: shipgen/EvtCalcGenerator.cxx [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  fTree->GetEntry(fn);
+  fn++;
+  if (fn %100==0)  cout << "Info EvtCalcGenerator: event nr "<< fn << endl;
+
+  //Vertex coordinates in the SHiP reference frame, expressed in [cm]
+  Double_t space_unit_conv = 100.;            // m to cm
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:92]: Can you use the `m` constant from `ShipUnit.h` instead?

File: shipgen/EvtCalcGenerator.cxx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  if (fn %100==0)  cout << "Info EvtCalcGenerator: event nr "<< fn << endl;
+
+  //Vertex coordinates in the SHiP reference frame, expressed in [cm]
+  Double_t space_unit_conv = 100.;            // m to cm 
+  Double_t coord_shift     = (zDecayVolume - ztarget)/space_unit_conv; // units m
+  Double_t vx_transf = vx * space_unit_conv; // units cm
+  Double_t vy_transf = vy * space_unit_conv; // units cm
+  Double_t vz_transf = (vz - coord_shift) * space_unit_conv; // units cm
+
+  Double_t c = 2.99792458e+10; // speed of light [cm/s]
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:98]: It would be good to use the one provided by `ShipUnit.h`

File: shipgen/EvtCalcGenerator.cxx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  Double_t tof = TMath::Sqrt(vx_transf*vx_transf + vy_transf*vy_transf 
+                            + vz_transf*vz_transf)/c;
+  
+// Mother LLP
+  Bool_t wanttracking=false;
+  pdg_llp = 0.; // Geantino, placeholder
+  cpg->AddTrack((Int_t)pdg_llp, px_llp, py_llp, pz_llp, vx_transf, vy_transf, vz_transf, -1.,
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:105]: Why not define the PDG codes as ints in the first place and avoid this conversion?

File: shipgen/EvtCalcGenerator.cxx [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  Double_t vy_transf = vy * space_unit_conv; // units cm
+  Double_t vz_transf = (vz - coord_shift) * space_unit_conv; // units cm
+
+  Double_t c = 2.99792458e+10; // speed of light [cm/s]
+  Double_t tof = TMath::Sqrt(vx_transf*vx_transf + vy_transf*vy_transf 
+                            + vz_transf*vz_transf)/c;
+  
+// Mother LLP
+  Bool_t wanttracking=false;
+  pdg_llp = 0.; // Geantino, placeholder
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:104]: Bad idea to use 0 here, which is by the way not the geantino (should be 998/999).

Better to follow the [standard numbering scheme](https://pdg.lbl.gov/2024/reviews/rpp2024-rev-monte-carlo-numbering.pdf) (see Section 11) or use the placeholder 99.

File: shipgen/EvtCalcGenerator.cxx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+#include "FairPrimaryGenerator.h"
+#include "EvtCalcGenerator.h"
+#include "TDatabasePDG.h"
+#include "TMath.h"
+#include "TGeoVolume.h"
+#include "TGeoNode.h"
+#include "TGeoManager.h"
+#include "TGeoEltu.h"
+#include "TVectorD.h"
+#include "TGeoCompositeShape.h"
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:13]: I find it hard to believe that all of these are used.

File: shipgen/EvtCalcGenerator.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  Double_t pdg_llp, pdg_prod1, pdg_prod2, pdg_prod3;
+  Double_t px_llp, py_llp, pz_llp, e_llp;
+  Double_t vx, vy, vz, decay_prob;
+  Double_t px_prod1, py_prod1, pz_prod1, e_prod1;
+  Double_t px_prod2, py_prod2, pz_prod2, e_prod2;
+  Double_t px_prod3, py_prod3, pz_prod3, e_prod3;
+  Double_t ztarget, zDecayVolume;
+  TFile* fInputFile;
+  TTree* fTree;
+  FairLogger*  fLogger; //!   don't make it persistent, magic ROOT command
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.h:48]: No longer needed.

File: shipgen/EvtCalcGenerator.h [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+ protected:
+  Double_t pdg_llp, pdg_prod1, pdg_prod2, pdg_prod3;
+  Double_t px_llp, py_llp, pz_llp, e_llp;
+  Double_t vx, vy, vz, decay_prob;
+  Double_t px_prod1, py_prod1, pz_prod1, e_prod1;
+  Double_t px_prod2, py_prod2, pz_prod2, e_prod2;
+  Double_t px_prod3, py_prod3, pz_prod3, e_prod3;
+  Double_t ztarget, zDecayVolume;
+  TFile* fInputFile;
+  TTree* fTree;
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.h:47]: Might be good to use smart-pointers for both of these.

File: test.txt
Comment (olantwin) [inline test.txt:1]: What's up with this file?

File: shipgen/EvtCalcGenerator.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
+#ifndef EVTCALCGENERATOR_H
+#define EVTCALCGENERATOR_H 1
+
+#include "TROOT.h"
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.h:4]: Why is this included?

File: macro/run_simScript.py [hunk 14]
```diff
@@ -380,6 +383,23 @@
...
+
+# -----EvtCalc--------------------------------------
+if simEngine == "EvtCalc":
+ print(f'Opening input file for conversion: {inputFile}')
+ if not os.path.isfile(inputFile): raise FileNotFoundError("EvtCalc: input .dat file does not exist")
+ primGen.SetTarget(0., 0.)
+ inputROOTFile = (convertEvtCalc.convert_file(infile=inputFile, outdir=options.outputDir))
+ print('Input root file: ', inputROOTFile)
+ print(f'Opening input file for EvtCalc generator: {inputROOTFile}')
+ ut.checkFileExists(inputROOTFile)
```
Comment (olantwin) [inline macro/run_simScript.py:395]: Would be better to do the conversion in a separate utility, and provide the ROOT file as input to run_simScript.py

File: shipgen/EvtCalcGenerator.h [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  Bool_t ReadEvent(FairPrimaryGenerator*);
+  virtual Bool_t Init(const char*, int); //!
+  virtual Bool_t Init(const char*); //!
+
+  Int_t GetNevents()
+  {
+    return fNevents;
+  }
+
+  void SetPositions(Double_t zTa=-5814.25, Double_t zDV=0.){
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.h:31]: I'd prefer not having hard-coded default arguments here.

**Review by olantwin — CHANGES_REQUESTED** [review]
First pass just reading over the code.

For testing this, could you please provide some example input files and instructions on how to use it?

**Comment by olantwin** [comment]
At the software meeting it was agreed that the todo items can be tackled later. @mferril, could you please open issue for them?

One thing that would be very nice to have would be some control plots for the expected output for some example channel and input file. Hopefully we can then gradually expand the collection to include more channels.

**Comment by olantwin** [comment]
@yabezsh @anupama-reghunath, did you have a chance to have a look at the pull request?

Overall, the proposed changes are quite small, so we should be able to merge this pretty quickly.

**Comment by mferril** [comment]
> @mferril @maksymovchynnikov Any updates?
> 
> > At the software meeting it was agreed that the todo items can be tackled later. @mferril, could you please open issues for them?

Hi @olantwin , I have given this lower priority given some conflicting deadlines this week. I am planning to look at it later in the day/weekend. Thanks for understanding!

**Comment by olantwin** [comment]
> Hi @olantwin , I have given this lower priority given some conflicting deadlines this week. I am planning to look at it later in the day/weekend. Thanks for understanding!

Ok, no worries! Good luck with the deadlines!

File: test.txt
Comment (mferril) [reply test.txt:1]: Solved

File: macro/run_simScript.py (hunk 14 above)
Comment (olantwin) [reply macro/run_simScript.py:395]: I'd prefer not having this is `run_simScript.py`. A nicer solution would be to add a small `if __name__ == "__main__":` section to `convertEvent.py` to allow it being called directly, allowing us to do the conversion for now.

And then in `run_simScript.py` we can use the output from that script.

Comment (mferril) [reply macro/run_simScript.py:395]: Solving for now. To be addressed later with a dedicated issue (interfacing EvtCalc sampling code).

File: shipgen/EvtCalcGenerator.cxx [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+using std::cout;
+using std::endl;
```
[the remark this answers is not in the document]
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:16]: Removed.

File: shipgen/EvtCalcGenerator.cxx (hunk 2 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:34]: Removed.

File: shipgen/EvtCalcGenerator.h (hunk 11 above)
Comment (olantwin) [reply shipgen/EvtCalcGenerator.h:48]: Unresolve until change is pushed.

File: macro/run_simScript.py (hunk 14 above)
Comment (olantwin) [reply macro/run_simScript.py:395]: Will look into this in separate issue #535

File: shipgen/EvtCalcGenerator.cxx (hunk 3 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx]: Removed.

File: shipgen/EvtCalcGenerator.cxx (hunk 4 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:30]: Removed.

File: shipgen/EvtCalcGenerator.cxx [hunk 17]
```diff
@@ -27,7 +27,7 @@ Bool_t EvtCalcGenerator::Init(const char* fileName, const int firstEvent) {
   fInputFile = TFile::Open(fileName, "read");
   LOGF(info, "Info EvtCalcGenerator: Opening input file %s", fileName);
   
-  fTree = (TTree *)fInputFile->Get("LLP_tree");
+  fTree = fInputFile->Get("LLP_tree");
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:30]: You will most likely need a `dynamic_cast<TTree*>(...)` here.

File: shipgen/EvtCalcGenerator.cxx (hunk 8 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:101]: The choice has been inherited from the conversion of the EvtCalc files to ROOT. 
I have changed that and Implemented it.

File: shipgen/EvtCalcGenerator.cxx (hunk 9 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx]: Implemented. I have left the Geantino placeholder for now. In the future it would be ideal to set up appropriate PDG codes within EvtCalc for each of the LLP.

File: shipgen/EvtCalcGenerator.cxx (hunk 10 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx]: Cleanup implemented.

File: shipgen/EvtCalcGenerator.h (hunk 11 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.h]: Commit pushed.

File: shipgen/EvtCalcGenerator.h (hunk 13 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.h]: Removed.

File: shipgen/EvtCalcGenerator.h (hunk 15 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.h]: Commit pushed.

**Comment by mferril** [comment]
> A preliminary remark and a question:
> 
> * The history needs some cleaning up, we can do that once everything else is ready for review. Only the merge commit might be nice to remove already now.
> * Which of the todo items would you consider necessary for the first version and which can we tackle later?

The items in the to-do list will be addressed in separate issues. The initial version of EventCalc can be merged without those enhancements.

**Comment by mferril** [comment]
> First pass just reading over the code.
> 
> For testing this, could you please provide some example input files and instructions on how to use it?

Agreed, I will provide some examples to run the vanilla implementation. It would be best if we do it once #535 is finalised.

File: macro/convertEvtCalc.py [hunk 18]
```diff
@@ -102,7 +102,10 @@ def convert_file(infile, outdir):
         branch = {}
         for var in vars_names:
             branch[var] = np.zeros(1, dtype=float)
-            tree.Branch(var, branch[var], f"{var}/D")
+            if 'pdg_' inbranch[var]:
```
Comment (olantwin) [inline macro/convertEvtCalc.py:105]: Typo

Comment (mferril) [reply macro/convertEvtCalc.py]: Fixed

File: shipgen/EvtCalcGenerator.h (hunk 12 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.h]: Implemented.

File: shipgen/EvtCalcGenerator.cxx [hunk 19]
```diff
@@ -17,10 +17,10 @@ Bool_t EvtCalcGenerator::Init(const char* fileName) {
 // -----   Default constructor   -------------------------------------------
 Bool_t EvtCalcGenerator::Init(const char* fileName, const int firstEvent) {
   
-  fInputFile = TFile::Open(fileName, "read");
+  fInputFile = std::make_unique<TFile>(TFile::Open(fileName, "read"));
```
Comment (olantwin) [inline shipgen/EvtCalcGenerator.cxx:20]: I didn't see this earlier, but this constructs a copy of the file, instead of adopting the pointer. Not a big issue, but I think just creating a unique_ptr would be better.

Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx]: Implemented.

File: shipgen/EvtCalcGenerator.cxx (hunk 6 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:82]: Opening a separate issue for this.

**Comment by olantwin** [comment]
@yabezsh @anupama-reghunath I'd really appreciate your feedback on your pull request, but will interpret continued silence as approval.

@mferril Could you please provide me a .dat file I can test this with? Apart from that this now looks to be in fairly good shape, and would be ready for automatic fixes, history rewriting etc.

**Comment by olantwin** [comment]
@mferril This also is still missing an entry in the changelog. I think an paragraph to introduce such a major new feature would be great, maybe with a link to your collaboration meeting presentation.

File: shipgen/EvtCalcGenerator.cxx (hunk 7 above)
Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:88]: Hi @olantwin , I haven't found time to complete this task yet. It seems that something is not adding up when using the Shipunits namespace. I am happy to proceed with the pull request without this enhancement if you think it is a good idea.

Comment (olantwin) [reply shipgen/EvtCalcGenerator.cxx:88]: That's fine, let's for now stick to the version that works.

Could you please open an issue for this? I think we keep redefining the speed of light in many many files, so we should try to fix this throughout the code base anyway.

Comment (mferril) [reply shipgen/EvtCalcGenerator.cxx:88]: It's already there: Issue #541

**Comment by mferril** [comment]
> @yabezsh @anupama-reghunath I'd really appreciate your feedback on your pull request, but will interpret continued silence as approval.
> 
> @mferril Could you please provide me a .dat file I can test this with? Apart from that this now looks to be in fairly good shape, and would be ready for automatic fixes, history rewriting etc.

I have just shared with you the input .dat file. To test whether it works, we could think about adding in the future some unit-testing or similar.

Basic commands to launch:
`python convertEvtCalc.py -f test_input.dat -o test_folder`
then
`python run_simScript.py --evtcalc -n 1 -o test_folder -f test_folder/test_input.root`

**Comment by mferril** [comment]
> @mferril This also is still missing an entry in the changelog. I think an paragraph to introduce such a major new feature would be great, maybe with a link to your collaboration meeting presentation.

@olantwin this has been changed in the last commit.

**Comment by olantwin** [comment]
> > @mferril This also is still missing an entry in the changelog. [..]
> 
> @olantwin this has been changed in the last commit.

Thanks! I'll take the liberty to slightly rephrase/reorganise it, but it provides the perfect amount of detail.

**Comment by olantwin** [comment]
> > @yabezsh @anupama-reghunath I'd really appreciate your feedback on your pull request, but will interpret continued silence as approval.
> > @mferril Could you please provide me a .dat file I can test this with? Apart from that this now looks to be in fairly good shape, and would be ready for automatic fixes, history rewriting etc.
> 
> I have just shared with you the input .dat file. To test whether it works, we could think about adding in the future some unit-testing or similar.
> 
> Basic commands to launch: `python convertEvtCalc.py -f test_input.dat -o test_folder` then `python run_simScript.py --evtcalc -n 1 -o test_folder -f test_folder/test_input.root`

Some observations:
1. When the output dir does not exist, the conversion script crashes. Might be worth creating it.
2. If the input file does not exist, the EventCalcGenerator crashes instead of exiting cleanly.

File: macro/convertEvtCalc.py
Comment (olantwin) [inline macro/convertEvtCalc.py:1]:
```
pydocstyle...............................................................Failed
- hook id: pydocstyle
- exit code: 1

macro/convertEvtCalc.py:12 in public function `parse_file`:
        D202: No blank lines allowed after function docstring (found 1)
macro/convertEvtCalc.py:12 in public function `parse_file`:
        D401: First line should be in imperative mood (perhaps 'Parse', not 'Parses')
macro/convertEvtCalc.py:75 in public function `convert_file`:
        D202: No blank lines allowed after function docstring (found 1)
macro/convertEvtCalc.py:75 in public function `convert_file`:
        D401: First line should be in imperative mood (perhaps 'Create', not 'Creates')
```

**Comment by yabezsh** [comment]
Hi @mferril ,
Thank you for the great work. I tested everything with the files provided and it worked smoothly on lxplus. The code looks good to me too.
I just have one ideological question. The first step is to convert the .dat file from EventCalc to the .root format, which will then be integrated into EventCalc. Will EventCalc be integrated into FairSHiP in the future, or will we use it as an external generator?
And the second question, the new physics particle now has pdg 999, do  we plan to have different pdg codes for different models? Or how can we distinguish in the generated files what kind of model has been simulated?
Thanks!

**Comment by olantwin** [comment]
> The first step is to convert the .dat file from EventCalc to the .root format, which will then be integrated into EventCalc. Will EventCalc be integrated into FairSHiP in the future, or will we use it as an external generator?

In my opinion EventCalc should be in a different repository, but ideally our simulation would call it directly without intermediate files. So close integration, and we could even consider maintaining/developing it in the collaboration, but I don't see any advantages and several disadvantages in including it within FairShip.

> And the second question, the new physics particle now has pdg 999, do we plan to have different pdg codes for different models? Or how can we distinguish in the generated files what kind of model has been simulated? Thanks!

This is an issue to be fixed after this initial version. For most, no standard PDG codes exist, and using the geantino as we do currently is no long-term solution. In the short-term, we need to differentiate via bookkeeping, but adding new particles to the PDG database similar to what we do for HNL would be better.

@yabezsh I interpret your message as approval?

**Comment by mferril** [comment]
> > The first step is to convert the .dat file from EventCalc to the .root format, which will then be integrated into EventCalc. Will EventCalc be integrated into FairSHiP in the future, or will we use it as an external generator?
> 
> In my opinion EventCalc should be in a different repository, but ideally our simulation would call it directly without intermediate files. So close integration, and we could even consider maintaining/developing it in the collaboration, but I don't see any advantages and several disadvantages in including it within FairShip.

@yabezsh this is currently under discussion and no final decision has been made yet on the integration within FairShip. We are working to achieve exactly what @olantwin mentioned, which is an internal call to the generator (e.g. as you would call Pythia or any other MC tool). See also Issue #537 .

> > And the second question, the new physics particle now has pdg 999, do we plan to have different pdg codes for different models? Or how can we distinguish in the generated files what kind of model has been simulated? Thanks!
> 
> This is an issue to be fixed after this initial version. For most, no standard PDG codes exist, and using the geantino as we do currently is no long-term solution. In the short-term, we need to differentiate via bookkeeping, but adding new particles to the PDG database similar to what we do for HNL would be better.
> 
Exactly, @olantwin explained it perfectly. Ideally we would issue separate PDG codes in the database for each LLP model we incorporate.

**Comment by yabezsh** [comment]
@mferril @olantwin ok, perfect. That's exactly what I meant, it would be good to be able to call EventCalc directly from FairSHiP so that the user doesn't have to search for .dat files or use EventCalc separately to generate files. Having it as a separate package is then much more practical.
then all is well from my point of  view, i approve

**Review by yabezsh — APPROVED** [review]

**Review by olantwin — APPROVED** [review]
